File: python/diffusion.py
#!/usr/bin/python3
import numpy as np
import math
# Nate Jackson
# Modeling the diffusion of a gas within a room
# Written for CSC 330 Project 2
# 11/18/2022

#user input
maxsize = int(input("Enter value for maxsize: "))
temp = input("75% partition y/n? ")

cube = np.zeros((maxsize+2, maxsize+2, maxsize+2)) #the room
diffusion_coefficient = 0.175
room_dimension = 5
speed_of_gas_molecules = 250.0
timestep = (room_dimension / speed_of_gas_molecules) / maxsize
distance_between_blocks = room_dimension / maxsize
DTerm = diffusion_coefficient * timestep / (distance_between_blocks * distance_between_blocks)
partition = False

mypass = 0
time = 0.0
ratio = 0.0
if temp == "y":
	partition = True
# determining location of partition
if partition:
	partX = int(math.floor((maxsize+1)/2))
	partY = int(math.ceil((maxsize+1)*0.25))
#placing partition
for i in range(1,maxsize+1):
	for j in range(1,maxsize+1):
		for k in range(1,maxsize+1):
			if (i == partX and j >= partY):
				cube[i][j][k] = 2.0

#first cell
cube[1][1][1] = 1.0e21

#function for diffusion between cells
def changing(i, j, k, l, m, n):
	if (( i == l ) and ( j == m ) and ( k == n+1)) or (( i == l ) and ( j == m ) and ( k == n-1) ) or ( ( i == l ) and ( j == m+1 ) and ( k == n)) or ( ( i == l ) and ( j == m-1 ) and ( k == n)) or ( ( i == l+1 ) and ( j == m )and ( k == n)) or ( ( i == l-1 ) and ( j == m ) and ( k == n)):
		if partition:
			if (cube[i][j][k] != 2.0) and (cube[l][m][n] != 2.0):
				change = (cube[i][j][k] - cube[l][m][n]) * DTerm
				cube[i][j][k] -= change
				cube[l][m][n] += change
		else:	
			change = (cube[i][j][k] - cube[l][m][n]) * DTerm
			cube[i][j][k] -= change
			cube[l][m][n] += change


while ratio < 0.99:
#call diffusion function across all cells
	[[[[[[changing(i,j,k,l,m,n) for n in range(1,maxsize+1) ] 
	for m in range(1, maxsize+1)]
	for l in range(1, maxsize+1)]
	for k in range(1, maxsize+1)]
	for j in range(1, maxsize+1)]
	for i in range(1, maxsize+1)]

	time += timestep
#check to see if room is fully diffused
	sumval = 0.0
	maxval = cube[1][1][1]
	minval = cube[1][1][1]
	for i in range(1,maxsize+1):
		for j in range(1,maxsize+1):
			for k in range(1,maxsize+1):
				if cube[i][j][k] != 2.0:
					maxval = max(cube[i][j][k], maxval)
					minval = min(cube[i][j][k], minval)
					sumval += cube[i][j][k]

	ratio = minval / maxval
#output
	print(time, " ", cube[1][1][1], " ", cube[maxsize][1][1], " ", cube[maxsize][maxsize][1], " ", cube[maxsize][maxsize][maxsize], " ", sumval)
	#end of while loop
print("Box equilibrated in ", time, " seconds of simulated time.")

File: julia/diffusion.jl
#!/usr2/local/julia-1.8.2/bin/julia
using Printf
# Nate Jackson
# Modeling the diffusion of a gas within a room
# For CSC 330 Project 2
# 11/18/2022

# User input
print("What is maxsize? ")
const maxsize = parse(Int32, readline())
print("75% parttition y/n? ")
temp = readline()

cube = zeros(maxsize+2 ,maxsize+2,maxsize+2) # the room
diffusion_coefficient = 0.175
room_dimension = 5.0
speed_of_gas_molecules = 250.0
timestep = (room_dimension / speed_of_gas_molecules) / maxsize
distance_between_blocks = room_dimension / maxsize
DTerm = diffusion_coefficient * timestep / (distance_between_blocks*distance_between_blocks)
p = false #partition
if temp == "y"
	p = true
end
#location of partition
partx = ceil((maxsize+2)/2)
party = ceil((maxsize+2)*0.5)
cube[2,2,2] = 1.0e21 # fill first cell

pass = 0
time = 0.0
ratio = 0.0
#creation of partition
if p
    for i in 1:maxsize+2, j in 1:maxsize+2, k in 1:maxsize+2
        if i == partx && j >= party
	        cube[i,j,k] = 2.0
        end
    end
end

while ratio < .99
    #diffusion
    for i in 2:maxsize+1, j in 2:maxsize+1, k in 2:maxsize+1, 
        l in 2:maxsize+1, m in 2:maxsize+1, n in 2:maxsize+1
        # check to see if cells are next to each other
        if i == l && j == m && k == n+1 ||
            i == l && j == m && k == n-1 ||
            i == l && j == m+1 && k == n ||
            i == l && j == m-1 && k == n ||
            i == l+1 && j == m && k == n ||
            i == l-1 && j == m && k == n
		if p
			if cube[i,j,k] != 2.0 && cube[l,m,n] != 2.0
        		change = (cube[i,j,k]- cube[l, m, n]) * DTerm
           		cube[i, j, k] -= change
           		cube[l,m,n] += change
			end
		end   
		if !p   
        	change = (cube[i,j,k]- cube[l, m, n]) * DTerm
           	cube[i, j, k] -= change
           	cube[l,m,n] += change
		end
        end
    end

    global time += timestep
    # check if room is fully diffused
    global sumval = 0.0
    global maxval = cube[2,2,2]
    global minval = cube[2,2,2]
    for i in 2:maxsize+1, j in 2:maxsize+1, k in 2:maxsize+1
	if cube[i,j,k] != 2.0
        global maxval = maximum([cube[i,j,k], maxval])
        global minval = minimum([cube[i,j,k], minval])
        global sumval += cube[i,j,k]
	end
    end
    global ratio = minval / maxval
    # Output
    @printf("%f %e",time, cube[2,2,2])
    @printf(" %e",cube[maxsize+1,2,2])
    @printf(" %e",cube[maxsize+1,maxsize+1,2])
    @printf(" %e",cube[maxsize+1,maxsize+1,maxsize+1])
    @printf(" %e\n",sumval)
end
@printf("Box equilibrated in %f seconds of simulated time.\n", time)
